source/2_platform/rex_directx/src/dxgi/adapter_manager.cpp

```diff
@@ -0,0 +1,154 @@
+
+namespace
+{
+  //-------------------------------------------------------------------------
+  rsl::function<HRESULT(UINT, IDXGIAdapter4**)> get_enumaration_function6(const rex::dxgi::Factory* factory)
+  {
+    const rex::wrl::com_ptr<IDXGIFactory6> factory_6 = factory->as<IDXGIFactory6>();
+
+    REX_ASSERT_X(factory_6, "DXGIFactory 6 does not exist!");
+
+    return [factory = factory_6.Get()](UINT index, IDXGIAdapter4** adapter) { return factory->EnumAdapterByGpuPreference(index, DXGI_GPU_PREFERENCE_HIGH_PERFORMANCE, IID_PPV_ARGS(adapter)); };
+  }
+  //-------------------------------------------------------------------------
+  rsl::function<HRESULT(UINT, IDXGIAdapter1**)> get_enumaration_function1(const rex::dxgi::Factory* factory)
+  {
+    const rex::wrl::com_ptr<IDXGIFactory4> factory_4 = factory->as<IDXGIFactory4>();
+
+    REX_ASSERT_X(factory_4, "DXGIFactory 4 does not exist!");
+
+    return [factory = factory_4.Get()](UINT index, IDXGIAdapter1** adapter) { return factory->EnumAdapters1(index, adapter); };
+  }
+  //-------------------------------------------------------------------------
+  rsl::function<HRESULT(UINT, IDXGIAdapter**)> get_enumaration_function(const rex::dxgi::Factory* factory)
+  {
+    REX_ASSERT_X(factory, "DXGIFactory does not exist");
+
+    return [factory = factory->c_ptr()](UINT index, IDXGIAdapter** adapter) { return factory->EnumAdapters(index, adapter); };
+  }
+
+  //-------------------------------------------------------------------------
+  template <typename DXGIAdapterInterface>
+  rsl::vector<rsl::unique_ptr<rex::Gpu>> get_adapters(const rsl::function<HRESULT(UINT, DXGIAdapterInterface**)>& enumarationFnc, uint32 version)
+  {
+    uint32 i                      = 0;
+    DXGIAdapterInterface* adapter = nullptr;
+
+    rsl::vector<rsl::unique_ptr<rex::Gpu>> adapters;
+    while(enumarationFnc(i, &adapter) != DXGI_ERROR_NOT_FOUND)
+    {
+      if(adapter)
+        adapters.push_back(rsl::make_unique<rex::dxgi::Adapter>(adapter, version));
+
+      ++i;
+    }
+
+    return adapters;
+  }
+
+  //-------------------------------------------------------------------------
+  rsl::unique_ptr<rex::Gpu> get_software_adapter(const rex::dxgi::Factory* factory)
+  {
+    REX_ASSERT_X(factory, "DXGIFactory does not exist");
+
+    if(factory->version() < 4)
+    {
+      REX_ERROR(LogDirectX, "Cannot create software rasterizer, incompatible driver");
+      return nullptr;
+    }
+
+    IDXGIAdapter* adapter = nullptr;
+
+    if(SUCCEEDED(factory->as<IDXGIFactory4>()->EnumWarpAdapter(IID_PPV_ARGS(&adapter))))
+    {
+      return rsl::make_unique<rex::dxgi::Adapter>(adapter, -1);
+    }
+
+    return nullptr;
+  }
+} // namespace
+
+namespace rex
+{
+  namespace dxgi
+  {
+    //-------------------------------------------------------------------------
+    AdapterManager::AdapterManager(const Factory* factory, const GpuScorer& scorer)
+        : m_selected_adapter(nullptr)
+    {
+      load_adapters(factory);
```

Comment:
shouldn't this assert that you can always load at least 1 (load_adapters returns true if any adapters are loaded)